SU2_CFD/src/drivers/CDriverBase.cpp

```diff
@@ -0,0 +1,756 @@
+  const auto nVertex = GetNumberMarkerNodes(iMarker);
+  unsigned long nHalo = 0;
+
+  for (auto iVertex = 0ul; iVertex < nVertex; iVertex++) {
+    auto iPoint = GetMarkerVertices(iMarker, iVertex);
+
+    if (!(main_geometry->nodes->GetDomain(iPoint))) {
+      nHalo += 1;
+    }
+  }
+
+  return nHalo;
+}
+
+vector<unsigned long> CDriverBase::GetMarkerVertices(unsigned short iMarker) const {
+  const auto nVertex = GetNumberMarkerNodes(iMarker);
+
+  vector<unsigned long> values;
+
+  for (auto iVertex = 0ul; iVertex < nVertex; iVertex++) {
+    values.push_back(GetMarkerVertices(iMarker, iVertex));
+  }
+
+  return values;
+}
+
+unsigned long CDriverBase::GetMarkerVertices(unsigned short iMarker, unsigned long iVertex) const {
+  if (iVertex >= GetNumberMarkerNodes(iMarker)) {
+    SU2_MPI::Error("Vertex index exceeds marker size.", CURRENT_FUNCTION);
+  }
+
+  return geometry_container[MESH_0][INST_0][ZONE_0]->vertex[iMarker][iVertex]->GetNode();
+}
+
+vector<unsigned long> CDriverBase::GetNodes() const {
+  const auto nPoint = GetNumberNodes();
+
+  vector<unsigned long> values;
+
+  for (auto iPoint = 0ul; iPoint < nPoint; iPoint++) {
+    values.push_back(GetNodes(iPoint));
+  }
+
+  return values;
+}
+
+unsigned long CDriverBase::GetNodes(unsigned long iPoint) const {
+  if (iPoint >= GetNumberNodes()) {
+    SU2_MPI::Error("Node index exceeds mesh size.", CURRENT_FUNCTION);
+  }
+
+  return main_geometry->nodes->GetGlobalIndex(iPoint);
+}
+
+vector<unsigned long> CDriverBase::GetMarkerNodes(unsigned short iMarker) const {
+  const auto nVertex = GetNumberMarkerNodes(iMarker);
+
+  vector<unsigned long> values;
+
+  for (auto iVertex = 0ul; iVertex < nVertex; iVertex++) {
+    values.push_back(GetMarkerNodes(iMarker, iVertex));
+  }
+
+  return values;
+}
+
+unsigned long CDriverBase::GetMarkerNodes(unsigned short iMarker, unsigned long iVertex) const {
+  auto iPoint = GetMarkerVertices(iMarker, iVertex);
+
+  return main_geometry->nodes->GetGlobalIndex(iPoint);
+}
+
+vector<bool> CDriverBase::GetDomain() const {
+  const auto nPoint = GetNumberNodes();
+
+  vector<bool> values;
+
+  for (auto iPoint = 0ul; iPoint < nPoint; iPoint++) {
+    values.push_back(GetDomain(iPoint));
+  }
+
+  return values;
+}
+
+bool CDriverBase::GetDomain(unsigned long iPoint) const {
+  if (iPoint >= GetNumberNodes()) {
+    SU2_MPI::Error("Node index exceeds mesh size.", CURRENT_FUNCTION);
+  }
+
+  return main_geometry->nodes->GetDomain(iPoint);
+}
+
+vector<bool> CDriverBase::GetMarkerDomain(unsigned short iMarker) const {
+  const auto nVertex = GetNumberMarkerNodes(iMarker);
+
+  vector<bool> values;
+
+  for (auto iVertex = 0ul; iVertex < nVertex; iVertex++) {
+    values.push_back(GetMarkerDomain(iMarker, iVertex));
+  }
+
+  return values;
+}
+
+bool CDriverBase::GetMarkerDomain(unsigned short iMarker, unsigned long iVertex) const {
+  auto iPoint = GetMarkerVertices(iMarker, iVertex);
+
+  return main_geometry->nodes->GetDomain(iPoint);
+}
```

Comment:
This API has too much boilerplate.
Either implement one function that returns a vector, or two functions, count + point-wise access.
And please just **mirror the c++ interface**...
For example, if you return the nodes of elements as global indices how do you read the solution at those nodes?
You are trying to make things too simple and convenient, which is good, except then it only works for one or two applications, exactly like our optimization scripts!
Do not apply additional transformations to the connectivity (domain bool and global index) which can otherwise be done as a single line in python... Because in doing so you automatically need to write one version of the function per transformation you want to offer to users, for every function!
Which results in too much code, and less functionality.
Just mirror the parts of the c++ interface that you think are useful, without adding more boilerplate to them, which can be expressed in single lines of python...

Reply:
I will write some thoughts here but we can bring this up for discussion at a dev meeting.

The new Python API requirements emerged from efforts to couple SU2 with external solvers in MDAO frameworks. The functions implemented in this PR and to a larger extent in PR #1750 (which I consider to be part 2 of the Python API refactoring) were necessary for those coupling efforts. These new functions were also found to be critical for standalone SU2 Python use, like performing residual-based adjoint sensitivity verification, geometry manipulations, and parameter sweeps for far-field conditions

Our experience through two independent research efforts led to specific requirements for the Python interface:
- Enable easy in-memory integration of primal and adjoint solvers.
- Increased modularity with a highly flexible API for execution and queries.
- Reduce startup development costs for future Python-based efforts. These include the creation of modular interfaces allowing SU2 to be driven by another framework and providing access to appropriate data structure in a useable format.
- Intuitive interface. Facilitate searchability of functions without in-depth experience with SU2 and its source code. This is also why some function names have been deliberately chosen to not directly mirror the C++ interface. We gave careful consideration for the standardization of the function naming towards this goal.

These requirements shaped the current Python API. Yes there is boilerplate code, but we think the convenience and the flexibility that is provided to users for a variety of applications is worth it

Reply:
> For example, if you return the nodes of elements as global indices how do you read the solution at those nodes? 

PR #1750 has new functions in python_wrapper_stucture.cpp that return the solution. There are two overloaded variations of the same function called `GetStates()`, and one of them takes an input argument as the global node index

Reply:
> Do not apply additional transformations to the connectivity (domain bool and global index) which can otherwise be done as a single line in python... Because in doing so you automatically need to write one version of the function per transformation you want to offer to users, for every function! Which results in too much code, and less functionality.

Not sure what you mean by the additional transformations. Can you please elaborate?

Even if it can be done in one line of Python, the whole motivation behind these API updates is to remove that obstacle for users. If they want connectivities, they will get them in a suitable data format directly from the API. We found these new functions to be vital for our research applications and other SU2 Python users should find them to be useful as well